File: crates/sc-auth/src/cred_cache.rs
//! DAV Basic-auth verification cache — tiers ① and ② of the verification
//! path.
//!
//! Two caches live here:
//! - `CredCache`: HMAC(ephemeral_key, "dav\0"‖user‖"\0"‖pw) -> Argon2 outcome,
//!   so a keep-alive-less client re-authenticating with the *same* password
//!   doesn't re-run Argon2. Positive TTL 15m absolute / 5m idle, negative 30s.
//! - `ConnMemo`: sha256(raw Authorization header bytes) -> Principal, meant
//!   to be held by the HTTP layer per-connection. Invalidated purely by the
//!   `auth_generation` counter (no separate TTL — a real per-connection memo
//!   dies with the connection; here it is a bounded LRU keyed by header hash
//!   instead, since this crate has no connection object to hang state off).
//! - `TokenCache`: sha256(app-password token) -> Principal, TTL 60s,
//!   also invalidated by `auth_generation`.

use crate::Principal;
use hmac::{Hmac, Mac};
use lru::LruCache;
use sha2::Sha256;
use std::num::NonZeroUsize;
use std::time::{Duration, Instant};

type HmacSha256 = Hmac<Sha256>;

#[derive(Clone)]
pub(crate) enum Outcome {
    Accepted(Principal),
    Rejected,
}

struct CredEntry {
    outcome: Outcome,
    gen: u64,
    inserted: Instant,
    last_hit: Instant,
}

pub(crate) struct CredCache {
    ephemeral_key: [u8; 32],
    map: parking_lot::Mutex<LruCache<[u8; 16], CredEntry>>,
    pos_ttl: Duration,
    pos_idle: Duration,
    neg_ttl: Duration,
}

impl CredCache {
    pub(crate) fn new(cap: usize, pos_ttl: Duration, pos_idle: Duration, neg_ttl: Duration) -> Self {
        let mut ephemeral_key = [0u8; 32];
        getrandom::getrandom(&mut ephemeral_key).expect("getrandom for ephemeral cred-cache key");
        Self {
            ephemeral_key,
            map: parking_lot::Mutex::new(LruCache::new(
                NonZeroUsize::new(cap.max(1)).unwrap(),
            )),
            pos_ttl,
            pos_idle,
            neg_ttl,
        }
    }

    pub(crate) fn key(&self, user: &str, pw: &str) -> [u8; 16] {
        let mut mac = HmacSha256::new_from_slice(&self.ephemeral_key).expect("hmac key any size");
        mac.update(b"dav\0");
        mac.update(user.as_bytes());
        mac.update(b"\0");
        mac.update(pw.as_bytes());
        let full = mac.finalize().into_bytes();
        let mut out = [0u8; 16];
        out.copy_from_slice(&full[..16]);
        out
    }

    /// Looks up a cache entry, validating generation + TTL. Returns `None`
    /// on miss or expiry (the caller must fall through to the rate gate +
    /// Argon2 path).
    pub(crate) fn get(&self, key: &[u8; 16], current_gen: u64) -> Option<Outcome> {
        let mut map = self.map.lock();
        let expired = {
            let entry = map.get(key)?;
            if entry.gen != current_gen {
                true
            } else {
                match &entry.outcome {
                    Outcome::Accepted(_) => {
                        entry.inserted.elapsed() > self.pos_ttl
                            || entry.last_hit.elapsed() > self.pos_idle
                    }
                    Outcome::Rejected => entry.inserted.elapsed() > self.neg_ttl,
                }
            }
        };
        if expired {
            map.pop(key);
            return None;
        }
        let entry = map.get_mut(key)?;
        entry.last_hit = Instant::now();
        Some(entry.outcome.clone())
    }

    pub(crate) fn put(&self, key: [u8; 16], outcome: Outcome, gen: u64) {
        let now = Instant::now();
        self.map.lock().put(
            key,
            CredEntry {
                outcome,
                gen,
                inserted: now,
                last_hit: now,
            },
        );
    }
}

struct ConnEntry {
    principal: Principal,
    gen: u64,
}

/// See module docs — this is a process-wide stand-in for the per-connection
/// memo described in ①. The HTTP layer is expected to key
/// it with `sha256(Authorization header bytes)`, exactly per the design.
pub(crate) struct ConnMemo {
    map: parking_lot::Mutex<LruCache<[u8; 32], ConnEntry>>,
}

impl ConnMemo {
    pub(crate) fn new(cap: usize) -> Self {
        Self {
            map: parking_lot::Mutex::new(LruCache::new(NonZeroUsize::new(cap.max(1)).unwrap())),
        }
    }

    pub(crate) fn check(&self, header_hash: &[u8; 32], current_gen: u64) -> Option<Principal> {
        let mut map = self.map.lock();
        let entry = map.get(header_hash)?;
        if entry.gen != current_gen {
            map.pop(header_hash);
            return None;
        }
        Some(entry.principal.clone())
    }

    pub(crate) fn store(&self, header_hash: [u8; 32], principal: Principal, gen: u64) {
        self.map.lock().put(header_hash, ConnEntry { principal, gen });
    }
}

struct TokenEntry {
    principal: Principal,
    gen: u64,
    inserted: Instant,
}

/// App-password verification cache: key = sha256(token),
/// TTL 60s, invalidated by `auth_generation` on revoke.
pub(crate) struct TokenCache {
    map: parking_lot::Mutex<LruCache<[u8; 32], TokenEntry>>,
    ttl: Duration,
}

impl TokenCache {
    pub(crate) fn new(cap: usize, ttl: Duration) -> Self {
        Self {
            map: parking_lot::Mutex::new(LruCache::new(NonZeroUsize::new(cap.max(1)).unwrap())),
            ttl,
        }
    }

    pub(crate) fn get(&self, token_hash: &[u8; 32], current_gen: u64) -> Option<Principal> {
        let mut map = self.map.lock();
        let expired = {
            let e = map.get(token_hash)?;
            e.gen != current_gen || e.inserted.elapsed() > self.ttl
        };
        if expired {
            map.pop(token_hash);
            return None;
        }
        map.get(token_hash).map(|e| e.principal.clone())
    }

    pub(crate) fn put(&self, token_hash: [u8; 32], principal: Principal, gen: u64) {
        self.map.lock().put(
            token_hash,
            TokenEntry {
                principal,
                gen,
                inserted: Instant::now(),
            },
        );
    }
}

File: go/internal/auth/credcache.go
package auth

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"sync"
	"time"

	"github.com/heavycaffeiner/stowcloud/go/internal/clock"
)

// Principal is what a successful credential resolves to, cached so a
// keep-alive-less client does not re-run Argon2 or a database lookup on every
// request.
type Principal struct {
	UserID   int64
	Display  string
	Disabled bool
}

// The three tiers of the verification path and their lifetimes, all of them
// invalidated together by the auth_generation counter. A password change, an
// app-password revocation or an account disable bumps that counter, so every
// entry in every tier goes stale by comparison rather than by being found and
// deleted: revocation is immediate on a surface that never re-reads the
// database.
const (
	// The tier-1 connection memo and the tier-3 app-password cache are LRUs
	// keyed by a digest with no TTL of their own beyond the generation counter.
	connMemoCapacity = 4096
	tokenCacheCap    = 1024

	// Tier 2: positive results live 15 minutes absolute, 5 minutes idle, so
	// an active sync stays warm and an abandoned client's entries evaporate.
	// Negative results live 30 seconds, because a client looping with a wrong
	// password would otherwise be a full Argon2 invocation on their behalf
	// and the server's, which is the denial of service arriving from the
	// direction nobody watches.
	credPositiveTTL  = 15 * time.Minute
	credPositiveIdle = 5 * time.Minute
	credNegativeTTL  = 30 * time.Second

	// App passwords are high-entropy and do not need a memory-hard function
	// to be safe, so a verified token bypasses tier 3 for a minute.
	tokenTTL = 60 * time.Second
)

// Outcome is a cached tier-2 or tier-3 decision.
type Outcome struct {
	Accepted  bool
	Principal Principal
}

type credEntry struct {
	outcome  Outcome
	gen      int64
	inserted time.Time
	lastHit  time.Time
}

type connEntry struct {
	principal Principal
	gen       int64
}

type tokenEntry struct {
	principal Principal
	gen       int64
	inserted  time.Time
}

// caches is the three tiers plus the ephemeral key that names tier-2 entries.
// All of them live and die with the process: the key comes from crypto/rand
// and is never on disk, so a process dump yields nothing offline-attackable.
type caches struct {
	ephemeral [32]byte

	credential *lru[[16]byte, credEntry]
	connMemo   *lru[[32]byte, connEntry]
	token      *lru[[32]byte, tokenEntry]

	clk clock.Clock
}

func newCaches(clk clock.Clock) *caches {
	var key [32]byte
	if _, err := rand.Read(key[:]); err != nil {
		// crypto/rand failing means the system RNG is gone; a cache that can
		// name its entries with a predictable key is worse than no cache.
		panic("crypto/rand failed to seed the credential cache: " + err.Error())
	}
	return &caches{
		ephemeral:  key,
		credential: newLRU[[16]byte, credEntry](4096),
		connMemo:   newLRU[[32]byte, connEntry](connMemoCapacity),
		token:      newLRU[[32]byte, tokenEntry](tokenCacheCap),
		clk:        clk,
	}
}

// tier2Key derives the cache key for one (user, password) pair: an HMAC under
// the per-process ephemeral key, never a hash of the password.
func (c *caches) Tier2Hash(user, pw string) [16]byte {
	input := make([]byte, 0, len(user)+len(pw)+2)
	input = append(input, 'd', 'a', 'v', 0)
	input = append(input, user...)
	input = append(input, 0)
	input = append(input, pw...)
	mac := hmac.New(sha256.New, c.ephemeral[:])
	mac.Write(input) //nolint:errcheck // hash.Hash.Write never fails.
	var out [16]byte
	copy(out[:], mac.Sum(nil)[:16])
	return out
}

// Tier2Lookup returns the cached outcome, or ok=false on a miss or an entry a
// generation or a TTL has made stale.
func (c *caches) Tier2Lookup(key [16]byte, gen int64) (Outcome, bool) {
	c.credential.mu.Lock()
	defer c.credential.mu.Unlock()
	e, present := c.credential.peek(key)
	if !present {
		return Outcome{}, false
	}
	now := c.clk.Now()
	switch {
	case e.gen != gen:
		c.credential.remove(key)
		return Outcome{}, false
	case e.outcome.Accepted:
		stale := now.Sub(e.inserted) > credPositiveTTL || now.Sub(e.lastHit) > credPositiveIdle
		if stale {
			c.credential.remove(key)
			return Outcome{}, false
		}
		e.lastHit = now
	default:
		if now.Sub(e.inserted) > credNegativeTTL {
			c.credential.remove(key)
			return Outcome{}, false
		}
	}
	e, _ = c.credential.peek(key)
	return e.outcome, true
}

func (c *caches) Tier2Store(key [16]byte, o Outcome, gen int64) {
	c.credential.mu.Lock()
	defer c.credential.mu.Unlock()
	now := c.clk.Now()
	c.credential.put(key, credEntry{outcome: o, gen: gen, inserted: now, lastHit: now})
}

// tier1Lookup is the connection memo: the resolved principal for a presented
// Authorization header, keyed by its SHA-256, invalidated by nothing but the
// generation counter.
func (c *caches) Tier1Lookup(hash [32]byte, gen int64) (Principal, bool) {
	c.connMemo.mu.Lock()
	defer c.connMemo.mu.Unlock()
	e, present := c.connMemo.peek(hash)
	if !present {
		return Principal{}, false
	}
	if e.gen != gen {
		c.connMemo.remove(hash)
		return Principal{}, false
	}
	return e.principal, true
}

func (c *caches) Tier1Store(hash [32]byte, p Principal, gen int64) {
	c.connMemo.mu.Lock()
	defer c.connMemo.mu.Unlock()
	c.connMemo.put(hash, connEntry{principal: p, gen: gen})
}

// tokenLookup is the app-password tier-3 bypass, keyed by sha256(token) with
// a 60-second TTL and the same generation invalidation as the other two
// tiers.
func (c *caches) TokenLookup(hash [32]byte, gen int64) (Principal, bool) {
	c.token.mu.Lock()
	defer c.token.mu.Unlock()
	e, present := c.token.peek(hash)
	if !present {
		return Principal{}, false
	}
	switch {
	case e.gen != gen:
		c.token.remove(hash)
		return Principal{}, false
	case c.clk.Now().Sub(e.inserted) > tokenTTL:
		c.token.remove(hash)
		return Principal{}, false
	}
	return e.principal, true
}

func (c *caches) TokenStore(hash [32]byte, p Principal, gen int64) {
	c.token.mu.Lock()
	defer c.token.mu.Unlock()
	c.token.put(hash, tokenEntry{principal: p, gen: gen, inserted: c.clk.Now()})
}

// lru is a bounded map with FIFO-evicted entries, guarded by its own mutex.
// Access-order eviction would take a list of links to do properly; FIFO keeps
// the bound that stops an attacker filling memory and is all a cache of
// verifiable credentials needs.
type lru[K comparable, V any] struct {
	mu   sync.Mutex
	cap  int
	data map[K]V
	keys []K
}

func newLRU[K comparable, V any](capacity int) *lru[K, V] {
	return &lru[K, V]{cap: capacity, data: map[K]V{}}
}

func (l *lru[K, V]) peek(k K) (V, bool) {
	v, ok := l.data[k]
	return v, ok
}

func (l *lru[K, V]) put(k K, v V) {
	if _, ok := l.data[k]; !ok {
		if len(l.keys) >= l.cap {
			old := l.keys[0]
			l.keys = l.keys[1:]
			delete(l.data, old)
		}
		l.keys = append(l.keys, k)
	}
	l.data[k] = v
}

func (l *lru[K, V]) remove(k K) {
	delete(l.data, k)
}
